webapp/src/serial.tsx

```diff
@@ -0,0 +1,309 @@
+    chartIdx: number = 0
+    consoleBuffer: string = ""
+    // TODO pass these values in with props or config
+    isSim: boolean = true
+    maxConsoleLineLength: number = 500
+    maxConsoleEntries: number = 100
+    active: boolean = true
+    rawDataBuffer: string = ""
+    //TODO reasonable buffer size?
+    maxBufferLength: number = 5000
+
+    getId() {
+        return "serialEditor"
+    }
+
+    acceptsFile(file: pkg.File) {
+        return file.name === pxt.SERIAL_EDITOR_FILE
+    }
+
+    isGraphable(v: string) {
+        return /[a-z]*:[0-9.]*/.test(v)
+    }
+
+    setSim(b: boolean) {
+        this.isSim = b
+        this.clear()
+    }
+
+    constructor(public parent: pxt.editor.IProjectView) {
+        super(parent)
+        window.addEventListener("message", this.processMessage.bind(this), false)
+    }
+
+    processMessage(ev: MessageEvent) {
+        let msg = ev.data
+        if (this.active && msg.type === "serial") {
+            //TODO y tho
+            const smsg = msg as pxsim.SimulatorSerialMessage
+            const sim = !!smsg.sim
+            if (sim == this.isSim) {
+                const data = smsg.data || ""
+                const source = smsg.id || "?"
+                //TODO not using theme anymore
+                let theme = source.split("-")[0] || "black"
+
+                //TODO incorporate source?
+                this.appendRawData(data)
+
+                if (this.isGraphable(data)) {
+                    this.appendGraphEntry(source, data, theme, sim)
+                } else {
+                    //TODO incorporate source?
+                    this.appendConsoleEntry(data)
+                }
+            }
+        }
+    }
+
+    appendRawData(data: string) {
+        this.rawDataBuffer += data
+        let excessChars = this.rawDataBuffer.length - this.maxBufferLength 
+        if (excessChars > 0) {
+            this.rawDataBuffer = this.rawDataBuffer.slice(excessChars)
+        }
+    }
+
+    appendGraphEntry(source: string, data: string, theme: string, sim: boolean) {
+        let m = /^\s*(([^:]+):)?\s*(-?\d+)/i.exec(data)
+        let variable = m ? (m[2] || ' ') : undefined
+        let nvalue = m ? parseInt(m[3]) : null
+
+        //See if there is a "home chart" that this point belongs to -
+        //if not, create a new chart
+        let homeChart: Chart = undefined
+        for (let i = 0; i < this.charts.length; ++i) {
+            let chart = this.charts[i]
+            if (chart.shouldContain(source, variable)) {
+                homeChart = chart
+                break
+            }
+        }
+        if (homeChart) {
+            homeChart.addPoint(nvalue)
+        } else {
+            let newChart = new Chart(source, variable, nvalue, this.chartIdx)
+            this.chartIdx++
+            this.charts.push(newChart)
+            let serialChartRoot = document.getElementById("serialCharts")
+            serialChartRoot.appendChild(newChart.getElement())
```

Comment:
Should you add the point here?

Reply:
The point is actually added in the Chart constructor.  Post-beta TODO: make this clearer

Reply:
This is "code cleanup" https://github.com/Microsoft/pxt/issues/3113